### PR TITLE
[amdgpu-arch] Replace use of HSA with reading sysfs directly

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
For Linux systems, we currently use the HSA library to determine the
installed GPUs. However, this isn't really necessary and adds a
dependency on the HSA runtime as well as a lot of overhead. Instead,
this patch uses the `sysfs` interface exposed by `amdkfd` to do this
directly.

### CODE REVIEW

File: clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    ErrorOr<std::unique_ptr<MemoryBuffer>> BufferOrErr =
+        MemoryBuffer::getFileOrSTDIN(Path);
+    if (std::error_code EC = BufferOrErr.getError())
+      return 1;
+
+    long GFXVersion = 0;
+    for (line_iterator Lines(**BufferOrErr, false); !Lines.is_at_end();
+         ++Lines) {
+      if (Lines->starts_with("gfx_target_version")) {
+        if (Lines->drop_front(sizeof("gfx_target_version"))
```
Comment (arsenm) [inline clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp:54]: consume_front

**Comment by jhuber6** [comment]
> @jhuber6 can you comment on "lot of overhead" and if that matters? Also, not sure why the HSA library dependence is a problem. This seems to be exposing amdgpu-arch to more maintenance overhead.

Sometimes the driver will hang and since this is used inside of `clang` to support `--offload-arch=native` I've had cases where the compiler hangs forever, so I added a timeout to keep it from doing that. This removes that possibility.

**Comment by jhuber6** [comment]
> > @jhuber6 can you comment on "lot of overhead" and if that matters? Also, not sure why the HSA library dependence is a problem. This seems to be exposing amdgpu-arch to more maintenance overhead.
> 

Sometimes the driver will hang and since this is used inside of `clang` to support `--offload-arch=native` I've had cases where the compiler hangs forever, so I added a timeout to keep it from doing that. This removes that possibility. It's faster and since this will be installed on every single LLVM build, not everyone has ROCm so it would be nice for this to work.

**Comment by jhuber6** [comment]
> I don't really understand why cluster users are compiling on a system where the GPUs are being stressed, and I still don't see why it's a good idea to break layering for this case.

I'd like to eliminate a class of failures I've seen with `--offload-arch=native` either causing the compilation job to fail, hang forever, or take a really long time. I suppose I cannot attest to whether or not the Driver team will modify this interface, but since it's hard-coded in ROCT I would doubt it. This is also sufficiently tested by `libc` which uses `-mcpu=native` and has its own bot, along with some downstream OpenMP tests using `--offload-arch=native` so I don't think this would silently fail and make it into a release.

> Also, I wasn't aware that the "native" offload arch is supported by ROCm.

HIP does at least upstream, don't know if that was modified in ROCm.

**Comment by arsenm** [comment]
> I suppose I cannot attest to whether or not the Driver team will modify this interface,

They must never ever do this. It is their fault and problem if this breaks

**Comment by yxsamliu** [comment]
> > > @jhuber6 can you comment on "lot of overhead" and if that matters? Also, not sure why the HSA library dependence is a problem. This seems to be exposing amdgpu-arch to more maintenance overhead.
> > 
> > 
> > Sometimes the driver will hang and since this is used inside of `clang` to support `--offload-arch=native` I've had cases where the compiler hangs forever, so I added a timeout to keep it from doing that in the past. This removes that possibility entirely. I have also had reports from cluster users that it becomes very slow when others are stressing the GPU. It's faster and since this will be installed on every single LLVM build, not everyone has ROCm so it would be nice for this to work. I think ti's fair to do this as the fast-path on Linux systems and then fall-back to HIP if something goes terribly wrong.
> 
> I don't really understand why cluster users are compiling on a system where the GPUs are being stressed, and I still don't see why it's a good idea to break layering for this case. Also, I wasn't aware that the "native" offload arch is supported by ROCm.

The issue also happens to machines with one GPU if amdgpu-arch is executed multiple times in a short time due to some limitation of the driver. --offload-arch=native calls amdpu-arch to get the actual GPU archs.

File: clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  for (sys::fs::directory_iterator Begin(KFD_SYSFS_NODE_PATH, EC), End;
+       Begin != End; Begin.increment(EC)) {
+    if (EC)
+      return 1;
+
+    long Node = 0;
+    if (sys::path::stem(Begin->path()).consumeInteger(10, Node))
+      return 1;
+
+    SmallVector<char> Path(Begin->path().begin(), Begin->path().end());
```
Comment (arsenm) [inline clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp:44]: SmallString? I would also expect to not need the explicit begin/end and use the range constructor

File: clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+constexpr static long getMinor(long Ver) { return (Ver / 100) % 100; }
+constexpr static long getStep(long Ver) { return Ver % 100; }
+
+int printGPUsByKFD() {
+  SmallVector<std::pair<long, long>> Devices;
+  std::error_code EC;
+  for (sys::fs::directory_iterator Begin(KFD_SYSFS_NODE_PATH, EC), End;
+       Begin != End; Begin.increment(EC)) {
+    if (EC)
+      return 1;
```
Comment (shiltian) [inline clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp:38]: If just 1/0, use `bool` instead?

Comment (jhuber6) [reply clang/tools/amdgpu-arch/AMDGPUArchByKFD.cpp:38]: The interface currently uses `int`, so I figured I'd just keep it to minimize churn.

**Comment by b-sumner** [comment]
I feel like this is a workaround.  Can we not fix the "limitation of the driver" and if there is an issue with HSA overhead, shouldn't we file a ticket?

**Comment by jhuber6** [comment]
> I feel like this is a workaround. Can we not fix the "limitation of the driver" and if there is an issue with HSA overhead, shouldn't we file a ticket?

I don't know for sure, but I'd guess that the limitations seen may be related to the limit of 128 or so doorbells I think we have somewhere. I don't see this as a workaround, it deliberately gets the same information without requiring `/dev/kfd` to be opened. That's unlikely something we'd want to push into HSA.

**Comment by arsenm** [comment]
We shouldn't need to load a driver to query the devices on the system. This exists for this purpose, if there isn't going to be some ioctl to read it for you

**Comment by jhuber6** [comment]
Here's a question, should it respect `ROCR_VISIBLE_DEVICES`?

**Comment by shiltian** [comment]
> Here's a question, should it respect `ROCR_VISIBLE_DEVICES`?

I think it depends on whether you consider this as a ROCm toolchain. If not, I'd prefer not to be bound to any ROCm related stuff.

**Comment by jhuber6** [comment]
> > Here's a question, should it respect `ROCR_VISIBLE_DEVICES`?
> 
> I think it depends on whether you consider this as a ROCm toolchain. If not, I'd prefer not to be bound to any ROCm related stuff.

Agreed, maybe in the fork but makes sense to leave community LLVM alone.

**Comment by shiltian** [comment]
Maybe add this information into the beginning of the file.

**Comment by arsenm** [comment]
> > Here's a question, should it respect `ROCR_VISIBLE_DEVICES`?

No. That is a transient runtime debug flag. This is not executing code on the device

**Comment by b-sumner** [comment]
> We shouldn't need to load a driver to query the devices on the system. This exists for this purpose, if there isn't going to be some ioctl to read it for you

It depends on the device and the system.  In this case, since linux provides an alternative I agree it makes sense to use it, and now see why this is not a workaround.  Thanks.

**Comment by yxsamliu** [comment]
> I feel like this is a workaround. Can we not fix the "limitation of the driver" and if there is an issue with HSA overhead, shouldn't we file a ticket?

I think there was already a ticket but it cannot be fixed easily.

**Review by JonChesterfield — APPROVED** [review]
Absolutely yes, awesome. I had a todo to have the kernel export this under sysfs literally years ago and didn't get around to working out their commit structure, delighted to see it is exposed. The unreliable hsa calls has been a consistent nuisance for ages, great to see we don't need that hack any more.

Possible after-commit fix is to replace all the return 1 with different numbers (maybe an enum?) so that we have exit 0 on success and an exit code which indicates what went wrong on failure.

**Comment by JonChesterfield** [comment]
Oh. I now see there was a bunch of discussion about this, will add some context.

The driver has a hard limit on how many processes can open it at a time. clang calls this utility to ask what gpu to compile for by default. If you put those together, a parallel build on a vaguely modern desktop immediately blows through that process limit and proceeds to fail, so the user has to deliberately build code slowly or specify the gpu by hand, to work around our tooling falling over.

The limit on number of processes is generally reasonable - launching hundreds of processes that all open the GPU and allocate queues and launch kernels is generally a disaster and having the kernel return an equivalent to "too many open" is great. However in the specific case where we are only asking for trivial information, which doesn't need to allocate a queue or do anything whatsoever with the GPU, this is a spurious and annoying limitation.

I suppose it could be "fixed" in the driver - some sort of reference count which is incremented when you do something non-trivial instead of on open - but I'd expect the kernel people to tell us to stop opening hundreds of processes when none of them do any work.

One clumsy outstanding thing is that this should now be some code in a header that clang includes so that instead of the subprocess shell to handle arch=native clang just looks up the information directly.

@b-sumner does the additional context make the design choice clear?

**Comment by JonChesterfield** [comment]
> Would only work for Linux unfortunately, unless some Windows driver developer out there knows if there's some similar win32 magic.

Windows getting subprocess calls until their driver catches up (or someone points out how to do this) seems fine to me. Linux people get a slightly faster clang, and maybe even one that prints useful error messages while it dies derived from the return code.

**Pull request merged**